File: PathSum/recursive.java
public boolean hasPathSum(TreeNode root, int sum) {
        if(root == null) return false;
        if(root.right==null && root.left == null && root.val == sum) return true;
        return hasPathSum(root.left,sum-root.val) || hasPathSum(root.right,sum-root.val);
}
//////
public boolean hasPathSum(TreeNode root, int sum) {
       if(root == null){return false;}
        if(root.left == null && root.right == null && sum == root.val)return true;
        return hasPathSum(root.left, sum - root.val) || hasPathSum(root.right, sum - root.val);
}

/////
public class Solution {
    public boolean hasPathSum(TreeNode root, int sum) {
        if(root==null)return false;
        sum = sum - root.val;
        if(root.left==null && root.right==null){
            if(sum == 0)return true;
            else return false;
        }
        return hasPathSum(root.left,sum) || hasPathSum(root.right,sum);


    }


}

File: PathSum/recursive.cpp
class Solution { public:

bool hasPathSum(TreeNode* root, int sum) {
    if(root == NULL ){
        return false;
    }
    if(sum == root->val && root->left== NULL && root->right == NULL){
        return true;
    }
    return hasPathSum(root->left, sum - root->val) || hasPathSum(root->right, sum - root->val);
}
};

///
/*
    bool hasPathSum(TreeNode* root, int sum) {
        if( !root ) return false;
        if( root->val == sum && root->left == NULL && root->right == NULL ) return true;
        return hasPathSum(root->left, sum - root->val) || hasPathSum(root->right, sum - root->val);
    }
*/
    bool hasPathSum(TreeNode* root, int sum) {
        if( !root ) return false;

        stack<pair<TreeNode*,int> > s;
        s.push(make_pair(root, sum));
        int val;
        TreeNode* p;

        while( !s.empty() ) {
            p = s.top().first;
            val = s.top().second;
            s.pop();
            if( p->left == NULL && p->right == NULL && p->val == val ) 
                return true;
            if( p->right ) 
                s.push(make_pair(p->right, val - p->val));
            if( p->left ) 
                s.push(make_pair(p->left, val - p->val));
        }
        return false;
    }
